ISPN-12810 marshalling and encoding title (#9106)
infinispan/infinispan: documentation/src/main/asciidoc/topics/con_protobuf_encoding.adoc

[id='protobuf-encoding_{context}']
= Protobuf cache encoding
Protocol Buffers (Protobuf) is a lightweight binary media type for structured data.

> As a cache encoding, Protobuf gives you excellent performance as well as interoperability between client applications in different programming languages and remote server endpoints.

Reviewer: I'm not sure the meaning of "remote server endpoints" would be clear to a user. I tried to reword this, to make it clearer but I'm not sure my attempt is any better ...

"Utilising protobuf as a cache encoding provides excellent performance and allows for interoperability between the {brandname} clients available in different programming languages for both the HotRod and REST endpoints."

The following example shows a Protobuf message that describes a `Person` object:

[source,json,options="nowrap",subs=attributes+]
----
include::protobuf/personalshopper.proto[]
----

[discrete]
== Interoperability

> Because it is language neutral, Protobuf encoding means {brandname} can handle requests from client applications written in Java, C{plusplus}, C#, Dart, Python, Go, and more.

Reviewer: I'm guessing you got `Dart` from the google docs? :laughing:. AFAIK it hasn't really taken off outside of Google, so I think we can remove that one.

Author: honestly that's an odd one. it must have slipped in there via google docs but the weird thing is that I added the `{plusplus}` attribute and remember looking at this. Dart def doesn't belong here anyway. will remove.

> {brandname} uses a ProtoStream library to encode caches with the `application/x-protostream` media type.
> This encoding is similar to `application/x-protobuf` but the two media types have some differences and are not interchangeable.

Reviewer 2: Last time I looked, there was not an "official" Protobuf media type, I've seen a handful of media types being used for protobuf, so better not to make promises here :smile: 
I will check again, but nothing prevents adding an alias to an eventual standard for protobuf in the future.

```suggestion
{brandname} uses the media type `application/x-protostream` for Protobuf encoded caches. 
```

Author: I've added this to the new PR.

//Downstream content
ifdef::downstream[]
[TIP]
====
Always use Protobuf encoding for integration with any Red Hat technology because it allows communication between applications and services.
====
endif::downstream[]

Protobuf also enables clients on different remote endpoints, Hot Rod or REST, to operate on the same data.

> Because it is backed by the REST API, you can access and work with Protobuf-encoded caches through {brandname} Console.

Reviewer: I'm not sure what you mean by "backed  by", maybe "fully supported"?

[NOTE]
====
{brandname} uses a ProtoStream library to encode caches with the `application/x-protostream` media type.

> This encoding is similar to `application/x-protobuf` but the two media types have some differences that mean they are not interchangeable.

Reviewer 3: Now here I have to ask the connaiseurs, why are x-protobuf and x-protostream different and not interchangeable?  I have no idea why we have invented x-protostream. x-protobuf whould have been enough. Any thoughts anyone ?

Author: @anistor it's never too late. maybe we can deprecate x-protostream in Infinispan 13 and start using x-protobuf instead?

Reviewer 4: @anistor I have suggested this addition because AFAIK it's not possible to write a value with protobuf-java and read it with protostream (or vice-versa). Perhaps I was wrong, but they are still not interchangeable in the sense that we don't recognize `application/x-protobuf`.

Author: @danberindei and @anistor Do you guys feel like we need to make any changes here or is this OK? I think it's a good level of detail for the reader but maybe the wording is slightly negative? In Korea there's this great colloquialism "same same but different". That's kind of the spirit I'm aiming for here, not sure if it comes across.

Reviewer: > I have suggested this addition because AFAIK it's not possible to write a value with protobuf-java and read it with protostream (or vice-versa).

Is this only the case if `WrappedMessage` has been used? I thought that if a user marshalls/unmarshalls data with Protostream via `ProtobufUtil.to/fromByteArray` that the two would be compatible.

Author: @danberindei and @ryanemerson TIL but I think we need to provide some more specific guidance on when to use those ProtoStream parameters with encodings. I can take note of this and will add it to the content plan for 13.

Reviewer 2: Not sure if the type parameter in the protostream media type affects the conversion (or at least I don't remember seeing it)

ifdef::downstream[]
You should always use Protobuf cache encoding with the `application/x-protostream` media type for integration with any Red Hat technology because it allows communication between applications and services.
endif::downstream[]

[discrete]
== Queries

{brandname} needs a structured representation of data in caches for fast and reliable queries.
To search caches with the Ickle query language, you register Protobuf schema that describe your objects.

[discrete]
== Custom types

{brandname} includes an implementation of the ProtoStream API with native support for frequently used types, including `String` and `Integer`.

> If you want to store custom types in your caches, use ProtoStream marshalling to generate and register serialization contexts with {brandname} so that it can marshall your objects.

Author: We do provide instructions for this. However this is a concept topic rather than a procedure, so a different type of content.
